extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/output/RetriableS3OutputStream.java

```
@@ -0,0 +1,432 @@
+    objectMetadata.setContentLength(resultsSize);
+    final UploadPartRequest uploadPartRequest = new UploadPartRequest()
+        .withUploadId(uploadId)
+        .withBucketName(bucket)
+        .withKey(key)
+        .withFile(chunk.file)
+        .withPartNumber(chunk.id)
+        .withPartSize(chunk.length());
+
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Pushing chunk [%s] to bucket[%s] and key[%s].", chunk, bucket, key);
+    }
+    UploadPartResult uploadResult = s3.uploadPart(uploadPartRequest);
+    return uploadResult.getPartETag();
+  }
+
+  @Override
+  public void close() throws IOException
+  {
+    if (closed) {
+      return;
+    }
+    closed = true;
+    Closer closer = Closer.create();
+
+    // Closeables are closed in LIFO order
+    closer.register(() -> {
+      // This should be emitted as a metric
+      LOG.info("Total push time: [%s] ms", pushStopwatch.elapsed(TimeUnit.MILLISECONDS));
+    });
+
+    closer.register(() -> org.apache.commons.io.FileUtils.forceDelete(chunkStorePath));
+
+    closer.register(() -> {
+      try {
+        if (resultsSize > 0 && isAllPushSucceeded()) {
+          RetryUtils.retry(
+              () -> s3.completeMultipartUpload(
+                  new CompleteMultipartUploadRequest(config.getBucket(), s3Key, uploadId, pushResults)
+              ),
+              S3Utils.S3RETRY,
+              config.getMaxTriesOnTransientError()
+          );
+        } else {
+          RetryUtils.retry(
+              () -> {
+                s3.abortMultipartUpload(new AbortMultipartUploadRequest(config.getBucket(), s3Key, uploadId));
+                return null;
+              },
+              S3Utils.S3RETRY,
+              config.getMaxTriesOnTransientError()
+          );
+        }
+      }
+      catch (Exception e) {
+        throw new IOException(e);
+      }
+    });
+
+    try (Closer ignored = closer) {
+      if (!error) {
+        pushCurrentChunk();
+      }
+    }
```

This seems somewhat odd. It would be clearer to push the current chunk first, and to explicitly call `closer.close()`. Or, if we want try-with-with-reasources (which seems overkill here), the try should include the closer creation.

Is the closer actually buying us anything? Any reason to not just do the steps, in forward order, in this method? Or, is the goal here to catch & suppress errors?